src/order/omega_complete_partial_order.lean

```diff
@@ -0,0 +1,522 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import data.pfun
+import order.category.Preorder
+import tactic.wlog
+import tactic.find_unused
+
+/-!
+# Omega Complete Partial Orders
+
+An omega-complete partial order is a partial order with a supremum
+operation on increasing sequences indexed by natural numbers (which we
+call `ωSup`). In this sense, it is strictly weaker than join complete
+semi-lattices as only ω-sized totally ordered sets have a supremum.
+
+The concept of an omega-complete partial order (ωCPO) is useful for the
+formalization of the semantics of programming languages. Its notion of
+supremum helps define the meaning of recursive procedures.
+
+## Main definitions
+
+ * class `omega_complete_partial_order`
+ * `ite`, `map`, `bind`, `seq` as continuous morphisms
+
+## Instances of `omega_complete_partial_order`
+
+ * `roption`
+ * every `complete_lattice`
+ * pi-types
+ * product types
+ * `monotone_hom`
+
+## References
+
+ * [G. Markowsky, *Chain-complete posets and directed sets with applications*, https://doi.org/10.1007/BF02485815][markowsky]
+ * [J. M. Cadiou and Zohar Manna, *Recursive definitions of partial functions and their computations.*, https://doi.org/10.1145/942580.807072][cadiou]
+-/
+
+universes u v
+
+local attribute [-simp] roption.bind_eq_bind roption.map_eq_map
+open_locale classical
+
+namespace preorder_hom
+
+variables (α : Type*) (β : Type*) {γ : Type*} {φ : Type*}
+variables [preorder α] [preorder β] [preorder γ] [preorder φ]
+
+instance : preorder (α →ₘ β) :=
+preorder.lift preorder_hom.to_fun
+
+variables {β γ}
+
+/-- the constant function as a monotone function -/
+@[simps]
+def const (f : β) : α →ₘ β :=
+{ to_fun := function.const _ f,
+  monotone := assume x y h, le_refl _}
+
+variables {α} {α' : Type*} {β' : Type*} [preorder α'] [preorder β']
+
+/-- the diagonal function as a monotone function -/
+@[simps]
+def prod.diag : α →ₘ (α × α) :=
+{ to_fun := λ x, (x,x),
+  monotone := λ x y h, ⟨h,h⟩ }
+
+/-- the `prod.map` function as a monotone function -/
+@[simps]
+def prod.map (f : α →ₘ β) (f' : α' →ₘ β') : (α × α') →ₘ (β × β') :=
+{ to_fun := prod.map f f',
+  monotone := λ ⟨x,x'⟩ ⟨y,y'⟩ ⟨h,h'⟩, ⟨f.monotone h,f'.monotone h'⟩ }
+
+/-- the `prod.fst` projection as a monotone function -/
+@[simps]
+def prod.fst : (α × β) →ₘ α :=
+{ to_fun := prod.fst,
+  monotone := λ ⟨x,x'⟩ ⟨y,y'⟩ ⟨h,h'⟩, h }
+
+/-- the `prod.snd` projection as a monotone function -/
+@[simps]
+def prod.snd : (α × β) →ₘ β :=
+{ to_fun := prod.snd,
+  monotone := λ ⟨x,x'⟩ ⟨y,y'⟩ ⟨h,h'⟩, h' }
+
+/-- the `prod` constructor as a monotone function -/
+@[simps {rhs_md := semireducible}]
+def prod.zip (f : α →ₘ β) (g : α →ₘ γ) : α →ₘ (β × γ) :=
+(prod.map f g).comp prod.diag
+
+/-- the `if _ then _ else _` function as a monotone function -/
+@[simps]
+def ite (p : Prop) [h : decidable p] (f g : α →ₘ β) :
```

Comment:
Isn't this just the same thing as `if p then f else g`? (Same for the omega complete version.) There doesn't seem to be any need for a new constructor. To make this interesting, `p` should be allowed to depend on `α` (in some continuous way, whatever that means).

Reply:
With `p` depending on the argument, except for a discrete order, we don't have a monotone function or a continuous function. With an additional assumption `f ≤ g`, it could be made monotone but that makes it a less interesting construct. 

Even though `ite` is not  a complicated definition, it is useful for constructing the fixed point of realistic functions so I'd like to keep it as is.

Reply:
My point is that the existing version, where `p` does not depend on the argument, can be easily expressed as `if p then f else g` (which is to say, you can replace the body of the definition with this and it will typecheck). You don't need a specialized version for continuous functions, and all the properties of it are just if-then-else case splits as well.